server/memo_comment.go

```diff
@@ -0,0 +1,191 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"github.com/usememos/memos/api"
+	apiv1 "github.com/usememos/memos/api/v1"
+	"github.com/usememos/memos/common"
+	"github.com/usememos/memos/store"
+)
+
+func (s *Server) registerMemoCommentRoutes(g *echo.Group) {
+	g.GET("/memo/comment/:memoId", func(c echo.Context) error {
+		ctx := c.Request().Context()
+		memoID, err := strconv.Atoi(c.Param("memoId"))
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("ID is not a number: %s", c.Param("memoId"))).SetInternal(err)
+		}
+
+		memoCommentsMessage, err := s.Store.GetMemoComments(ctx, &store.FindMemoCommentMessage{
+			MemoID: &memoID,
+		})
+		if err != nil {
+			if common.ErrorCode(err) == common.NotFound {
+				return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("Memo ID not found: %d", memoID)).SetInternal(err)
+			}
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to find memo by ID: %v", memoID)).SetInternal(err)
+		}
+
+		// userID, ok := c.Get(getUserIDContextKey()).(int)
+		// the map key is parent id , value is parent's reply, 0 is parent
+		mapComments := make(map[int][]*apiv1.MemoCommentResponse)
+		for i, message := range memoCommentsMessage {
+			if message.Visibility == store.Private {
+				continue
+			} else if message.Visibility == store.Protected {
+				memoCommentsMessage[i].Content = "Protected"
+			}
+			memoCommentResponse, err := s.composeMemoCommentMessageToMemoCommentResponse(ctx, message)
+			if err == nil {
+				if mapComments[memoCommentResponse.ParentID] == nil {
+					mapComments[memoCommentResponse.ParentID] = []*apiv1.MemoCommentResponse{
+						0: memoCommentResponse,
+					}
+				} else {
+					mapComments[memoCommentResponse.ParentID] = append(mapComments[memoCommentResponse.ParentID],
+						memoCommentResponse)
+				}
+			}
+		}
+
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to compose memo comment response").
+				SetInternal(err)
+		}
+		return c.JSON(http.StatusOK, composeResponse(mapComments))
+	})
+
+	g.POST("/memo/comment", func(c echo.Context) error {
```

Comment:
If it just about comment, `/comment` will be enough. If it is about comments of a memo, `/memo/:memoId/comment` will be better.

Reply:
ok,finished